### ISSUE TITLE
Discussion: Jitsi Integration

### ISSUE THREAD

**Issue by sunjam** [nextcloud/spreed#3173]
[Jitsi](https://jitsi.org/) Integration would be very useful since it is a [well established](https://jitsi.org/built-on-jitsi/), fully open source videoconferencing tool also built on WebRTC + XMPP #826 [as described here](https://xmpp.org/uses/webrtc.html) via [Videobridge](https://jitsi.org/jitsi-videobridge/).

Would be an excellent integration, which includes a number of improvements requested for Talk.  [Forum discussion on help.nextcloud here](https://help.nextcloud.com/t/jitsi-integration-in-nextcloud-talk/75445).
* Easily join or create a registration-free session on https://meet.jit.si/

Related issues
#328 #21 #1056 #2458 #3106 #3151 
[e2e encryption #37](https://github.com/nextcloud/spreed/issues/37) and [external api for sip bridging only offered in MCU #40](https://github.com/nextcloud/spreed/issues/40).
Is the MCU itself built with Jitsi?

[Jitsi Integration is already used by the Matrix project](https://jitsi.org/news/matrix-adds-support-for-video-with-jitsi/) #809 and Rocket.chat #753 

[Request for Caldav support in Jitsi](https://github.com/jitsi/jitsi-meet/issues/4719) and [request for Webdav support](https://github.com/jitsi/jitsi-meet/issues/5412)

**Comment by nickvergessen** [comment]
> Is the MCU itself built with Jitsi?

No

**Comment by asteven** [comment]
I would also love to have jitsi integration. Or better yet full matrix integration.
But I understand that the nextcloud folks also have to pay their bills somehow.
Why would they be interested to develop some jitsi integration that takes the butter of their bread?
They also have families to feed at home, like everybody else.

**Comment by warnerbryce** [comment]
> I would also love to have jitsi integration. Or better yet full matrix integration.
> But I understand that the nextcloud folks also have to pay their bills somehow.
> Why would they be interested to develop some jitsi integration that takes the butter of their bread?
> They also have families to feed at home, like everybody else.

Yes that why, someone (or a team) have to build another App for Matrix or Jitsi in the Nextcloud App Store. Nextcloud GmbH will have nothing to block this integration and if the App is really brillant, it will be well spread over the community. 
Don't forget that Nextcloud Talk cames from Spreed https://www.spreed.eu/ and so you can't rewrite it to add Jitsi. 
I think we should create a topic on help.nextcloud.com asking for Dev about Jitsi or Matrix for Nextcloud.

**Comment by sunjam** [comment]
> I think we should create a topic on help.nextcloud.com asking for Dev about Jitsi or Matrix for Nextcloud.

https://help.nextcloud.com/t/jitsi-integration-in-nextcloud-talk/75445

**Comment by EliterScripts** [comment]
People keep saying that Nextcloud is basically doing an “invented here” because they want to make money.
The truth is, they make money off of support.

Now, correct me if I’m wrong, but Jitsi doesn’t really seem like a team collaborative software. They really kick-butt at video conferencing, but they don’t really lack security, and confidentiality features that teams need. They’re also not a file sharing platform.

What Nextcloud needs to do, is take Jitsi, and make their own spin on it. Jitsi doesn’t have groups, users, file sharing, plugins, etc.. Nextcloud can still use Jitsi and come out on top.

I think the only problem is that Nextcloud is based on PHP, so it will be quite different to be running a non-web-based language on Nextcloud, which means if they decide to go with it, users can’t use hosting providers that hand them a web server with PHP capabilities.

But honestly, I think it doesn’t matter. These days, we have Docker, so Nextcloud could just hand out a Docker image for Nextcloud Jitsi (or whatever the name it).

**Comment by antiuser** [comment]
Yeah, im vote for this integration. We are already using Nextcloud self-hosted, and Jitsi self-hosted. It will be great idea to integrate Jitsi to Nextcloud for video conferencing.

**Comment by pojntfx** [comment]
Just wrote a bot: https://github.com/pojntfx/nextcloud-talk-jitsi-bot

Pretty simple to use; add the bot to the chat, type `#videochat` and get a link to a Jitsi meeting.

**Comment by pojntfx** [comment]
Just so you know; this is massively important now that COVID-19 is hitting schools and my fellow FLOSS comrades have to help massive amounts of people confused by Nextcloud Talk's lack of proper conference support.

**Comment by LorbusChris** [comment]
I agree that having an open source, highly performant MCU backend implementation for Talk is now more important than ever.

I think to make this work spreed will have to learn to speak the XMPP protocol as an alternative to the [custom external signaling API](https://github.com/nextcloud/spreed/blob/master/docs/standalone-signaling-api-v1.md) that is currently has.

It could then be taught to _talk_ to [jicofo](https://github.com/jitsi/jicofo) (signaling server) via e.g. Prosody and to make use of [jitsi-videobridge](https://github.com/jitsi/jitsi-videobridge) for WebRTC routing.

Bonus benefit is that these two of course already know how to talk to their Jitsi ecosystem siblings [jigasi](https://github.com/jitsi/jigasi) (SIP bridge) and [ice4j](https://github.com/jitsi/ice4j) (TURN/STUN/ICE).

In other words Nextcloud Talk would have to learn of Jitsi Meet's APIs and be able to replace it.

**Comment by asteven** [comment]
Not sure about the details, but maybe it would be better to concentrate on integrating with matrix for this.

**Comment by LorbusChris** [comment]
Ms there an MCU or SFU backend for Matrix? Otherwise XMPP looks like the most straight forward way to enable the rich Jitsi ecosystem for use here.

**Comment by EliterScripts** [comment]
I don't think Jitsi is very secure, as-is. As far as I can tell, ANYONE can go onto the website, create their own video conference, add 100 people, and totally take advantage of the service without permission from the website owner. I also don't see any administrative tools.

Nextcloud, you need to capitalize on this opportunity!

**Comment by warnerbryce** [comment]
@EliterScripts hey it’s the open format here
When you configure your own jitsi instance you can put a username/password for creating rooms.

**Comment by EliterScripts** [comment]
@warnerbryce For the whole server or just one particular meeting? If it's just for one particular meeting, then that is a security flaw.

**Comment by warnerbryce** [comment]
@EliterScripts yes for the whole server

**Comment by jospoortvliet** [comment]
> Just so you know; this is massively important now that COVID-19 is hitting schools and my fellow FLOSS comrades have to help massive amounts of people confused by Nextcloud Talk's lack of proper conference support.

If somebody wants to improve Jitsi integration, go ahead of course, that is great. But please keep in mind that of the 300-400K Nextcloud users, only a few thousand are able to install something like Jitsi. The rest doesn't have the skills for that. So unless you can rewrite Jitsi in PHP and Javascript, for most people it still won't be an option. It only helps companies and government agencies and schools with paid system administrators and paid hosting solutions.

Now that's nice, but I'd like to think that those could also pay for the development of the software they use. Any way, I [wrote a rant about that on my blog](https://blog.jospoortvliet.com/2020/04/rant-of-day-well-at-least-microsoft-is.html) so I'll leave that out here.

In any case, we rewrote spreed from Go to PHP because that makes it easy to install and use - and yes, that comes with limitations. But we are working to try and improve Talk so it can handle larger numbers of people in a single call - think 10-15, perhaps. Help with that is of course welcome, too, I think it is time better spend than integrating other products.

**Comment by asteven** [comment]
> > Just so you know; this is massively important now that COVID-19 is hitting schools and my fellow FLOSS comrades have to help massive amounts of people confused by Nextcloud Talk's lack of proper conference support.
> 
> If somebody wants to improve Jitsi integration, go ahead of course, that is great. But please keep in mind that of the 300-400K Nextcloud users, only a few thousand are able to install something like Jitsi. The rest doesn't have the skills for that. So unless you can rewrite Jitsi in PHP and Javascript, for most people it still won't be an option. It only helps companies and government agencies and schools with paid system administrators and paid hosting solutions.

You could of course argue that from the 300-400K Nextcloud users only a few thousand (maybe few 10 thousand) are able to install something like Nextcloud in the first place. All others are using some click-to-deploy-platform or some hosted offering.

> Now that's nice, but I'd like to think that those could also pay for the development of the software they use. Any way, I [wrote a rant about that on my blog](https://blog.jospoortvliet.com/2020/04/rant-of-day-well-at-least-microsoft-is.html) so I'll leave that out here.

That's one of the best things I've lately read. Thanks for posting that.

It indeed is very sad that the givie-me-everything-for-free-and-i-will-give-nothing-back attitude of people has ever been increasing since opensource became more main stream.
It used to be 'scratch your own itch and share' and 'make the world a better place together'. Now you have people demanding that you go write code to solve their problems for free :-/.

> In any case, we rewrote spreed from Go to PHP because that makes it easy to install and use - and yes, that comes with limitations. But we are working to try and improve Talk so it can handle larger numbers of people in a single call - think 10-15, perhaps. Help with that is of course welcome, too, I think it is time better spend than integrating other products.

To be honest I'm not sure this makes sense. You can not scale this far enough with just plain PHP.
Then again I understand that you have no interest to invest time into integrating yet more other tools that essentially just makes you loose even more money.

**Comment by fti7** [comment]
Sadly thats Fundamental Issue with Nextcloud. That "lets reinvent the wheel, but in PHP. So its easy so install." approach is imho an Dead End.
Even Owncloud recognized that (Check the Switch Statements from PHP to Go)

It started nice, but the User Requirements are changing. (Even more if you need to compare with propritary Solutions like GSuite/Office356.)
Also the way they are using/deploying the Software Stack.
Now its PHP/Cheap Webspace. In the Future it will be (as @asteven already mentioned) click-to-deploy Hosters or solutions like small Docker Hosts or Kubernetes.

New User Requirements will soon or later move it from PHP to a more Daemonized/Always-running Components.
Checkout the Onlyoffice/Collabora Plugin. Its not really possible to do it in PHP Only. (Performance and/or manpower to rewrite it in PHP (Would be stupid anyway)) (I dont want to talk about the big Onlyoffice Community Server Blob! :-( )

Nextcloud is also lacking of a good Delta/Incremental Sync Support because of these Limitations.
And there are a lot more lost Chances!

Beside of that general Rant, Nextcloud Talk doesnt make it better.
The "Reinvent the Chat Wheel in PHP" again. Propritary HPB. A lot of lost chances.
Why not investing time and money in Solutions like a Matrix compatible Chat Client? Jitsi Client/Integration?
For both Solutions there will be also free Community Servers as well - if you dont want to operate it by your own.

Maybe i just doesnt understand Nextclouds Long-Term Vision :-) 
I would really except from Nextcloud that they push non-propritary Standards like they did with ActivityPub etc. - But that doesnt seem to pay the bills unfortunately.

**Comment by DanScharon** [comment]
> Why not investing time and money in Solutions like a Matrix compatible Chat Client? Jitsi Client/Integration?

Regarding Chat there already is a [XMPP Client](https://apps.nextcloud.com/apps/ojsxc) which already supports 1 to 1 WebRTC-based Video-Chats. Via [lib-jitsi-meet](https://github.com/jitsi/lib-jitsi-meet) this could be extended to include Jitsi Meet. Open issue is here: https://github.com/jsxc/jsxc/issues/3

**Comment by sunjam** [comment]
> So unless you can rewrite Jitsi in PHP and Javascript, for most people it still won't be an option. It only helps companies and government agencies and schools with paid system administrators and paid hosting solutions.

I want to respectfully share a different perspective, because there is something here worth digging into: Jitsi reaches a very diverse group of people.  Like Nextcloud, it suffers "open source isn't as good" nonsense.  Wonderfully, it allows registration-free video and audio calls by absolutely anyone without needing technical skills once your server is up and running!  Imagine if Talk could make calls without either party needing to login as a default.

The beauty of it is there is almost nothing to learn for the non-technical person using Jitsi, plus it can scale up very easily.  We're in the process of testing it for our 100% volunteer Educational nonprofit.  Due to the Corona virus we've gone all online so now a videochat has gone from rarely having a couple people to having 20 or 40 people on mute just listening in the background with their video feed turned off, plus another dozen with video on.  It helps the average person just as much as it does some enterprise business.

> 
> In any case, we rewrote spreed from Go to PHP because that makes it easy to install and use - and yes, that comes with limitations. But we are working to try and improve Talk so it can handle larger numbers of people in a single call - think 10-15, perhaps. Help with that is of course welcome, too, I think it is time better spend than integrating other products.

[Thanks for posting more technical details on the forum](https://help.nextcloud.com/t/videoconference-performance/74137/33?u=just).  Make no mistake, we love Nextcloud!  You have a noteworthy connection to the home enthusiast market.  That is why we enthusiasts are excited about these possible integrations and projects.  They are great and help us all get more connected as little people spinning up a 1-click vps!  I feel confident that volunteer developers will emerge, because this is a big deal for people.  They want to be able to conference in larger groups badly, especially in places where we can no longer meet face-to-face or even leave our homes.  [Cool note: this bot for starting Jitsi calls in Talk has already emerged](https://github.com/pojntfx/nextcloud-talk-jitsi-bot).

**Comment by EliterScripts** [comment]
> It only helps companies and government agencies and schools with paid system administrators and paid hosting solutions.
> 
> Now that's nice, but I'd like to think that those could also pay for the development of the software they use.
@jospoortvliet I'm sorry. but I was going to pay Nextcloud, to help out with the project, because my job is to do online demos, to sell knives. But it just isn't worth the money! Your sales rep quoted me a price of "5250 Euro" for "50 U Nextcloud Talk Enterprise Subscription with HPB", when Nextcloud Talk can't even do simple things, like setting up a public call with two people, with the host not having an avatar. It just isn't worth it. I need to have the confidence that the video software is going to work, and that I'm focused on the meeting, and not on the meeting software. I just don't trust Nextcloud Talk with my clients or my coworkers. Not that it isn't secure, it just doesn't work 60% of the time.

You say you aren't reinventing the wheel, but you are.

Installing Docker and running a Docker image is so EASY. You assume that everyone doesn't have access to anything outside of PHP, but the thing is, a Digital Ocean droplet costs $5/mo, and is really easy to get started.

You could even take Mail-in-a-Box's approach ( https://mailinabox.email/ ), where they tell you to have a clean Ubuntu image, then run the command:
`curl -s https://mailinabox.email/setup.sh | sudo bash`
This installs the packages, and manages the whole system. We have lots of people that don't know how to run a Linux box on Mail-in-a-Box's forum, and it's secure, and isn't a headache.

Yes, Jitsi has a Docker image. Maybe you could run openssh-server on the Docker image, and have a Nextcloud plugin use PHP to interface with SSH on the Docker image for Jitsi.

**Comment by warnerbryce** [comment]
A lot of HOW TO Install Nextcloud like @enoch85 show how to start with Nextcloud with a vm or a docker. 
Adding Jitsi with a docker container isn’t a big deal for a lot of small customers because they rely on those one click made solution. 
So yes Jitsi integration in Nextcloud isn’t a big deal, a little nice App that interface with a Jitsi instance (docker or dedicated server) and everyone who need it can install it easily (vm - docker - script - HOW TO)

**Comment by EliterScripts** [comment]
I mean, hey, Collabora is using Docker. That's because it's easy to maintain, with the rest of the system, and install.

**Comment by EliterScripts** [comment]
I just found out that Jitsi doesn't even provide commerical support!

https://jitsi.org/user-faq/ under `Where can I get commercial support for Jitsi?`
> Neither the immediate Jitsi team or 8×8 provide commercial support for Jitsi. Jitsi does enjoy a large developer community with many development shops and individuals that provide support and commercial development services.

**Comment by jospoortvliet** [comment]
> > 
> 
> EDIT: Again, Nextcloud needs to capitalize on this! Include Jitsi integration into Nextcloud Server, then provide support! Money, money, money!

You wouldn't get what you want, sorry.

You want Jitsi because it scales.

Jitsi scales because it isn't a PHP app.

In that regard, Jitsi is a lot like Spreed, which was a Go app. Scalable, but: hard to install (not just one click in the nextcloud interface). That means 95% of our users were not able to install and use Spreed.

We wanted everyone to have easy video calling and chat, so we rewrote Spreed from Go to PHP. Now it is easy to install and perfect for 99% of our users. It is NOT scalable, however, so 1% of our users is unhappy. For them, we worked with the people behind the original Spreed (which you can still find [here](https://github.com/strukturag/nextcloud-spreedme)) to make the HPB.

I hope this helps understand why this integration would not make sense: Jitsi is not written in PHP so it can't be integrated. We could rewrite it, but then it loses its scalability and - well, we already did that, it is called 'Talk'.

Some basic integration, like being able to embet Jitsi as 'an app' and being able to share files in it - that can be done, but has to be done on the Jitsi side. Note that this is similar to our integration in Rocket.chat, which we supported - a GSOC student did most of the work. We can certainly do something similar for Jitsi, if there's a student interested in that. Or you ask their developers to do that, we'd happily have a Jitsi integration app in apps.nextcloud.com and promote it.

But this is useful for less than 1% of our users, so it is not a priority. I know it is important to YOU but for most private users, Talk is great and is getting better.

And I personally am part of that 99%, I even looked at installing Jitsi and gave up. It is way to complicated for a non-professional-system-administrator.

So it will never be useful for the vast majority of Nextcloud users, like Talk is, unless you rewrite Jitsi in PHP and at that point - you simply have a second Talk and people will ask integration with, I dunno, BigBlueButton... ;-)

Now I'm sure there's another 50 people who will comment here that they want Jitsi integration - great, but we have 300.000-400.000 Nextcloud servers on the web and most of them will not have Jitsi. So you're welcome to find somebody willing to build integration (not here, it is not a Spreed feature, it would be a Jitsi feature and a Jitsi app on apps.nextcloud.com) and you can use [bountysource.com](https://www.bountysource.com/teams/nextcloud) perhaps to find funding for it.

So I'll close this now, Jitsi integration would be a Jitsi feature, not a Nextcloud feature and certainly not a Nextcloud Talk feature.

**Comment by damienvancouver** [comment]
Sorry to necromance a closed thread, but this is the definitive one you find when searching for jitsi integration!

First off, Thank you to all the devs for your incredible hard work on NC in general and Spreed in particular! 

I run a hosting company that hosts mainly small 1-10 person orgs and we've been providing a lot of Jitsi to them.  We're also at the tail end of rollin gout Nextcloud as a service to these types of orgs, and that's why I need a solution for this to get a jitsi button in the nextcloud app.  

I think there's actually a _really_ simple solution to this that will satisfy the people needing Jitsi.

All that's _needed_ for Jitsi in nextcloud is:

- An "Open Jitsi" button on the taskbar
- ....that opens to a URL you configure in the app settings.

.. that's it.  we don't need integration any  deeper than that.   There isn't any point to deeper integration than that, because of the way Jitsi chose to do things.

Read on for more on that.. or skip to the last paragraph if TL;DR...

 I've been real world testing Jitsi extensively for months now and how Jitsi Meet works, (for any reading that aren't familiar), is that there is no security layer (ie. no user logins), no accounts, no logging, no anything except the group call experience part.  that's by design.  You hit the homepage and enter your username and pick a random chat room name (or use one saved in a cookie from before to revisit an old room).  Once the last participant leaves, it's gone forever.  Nothing is logged.  There's an ephemeral chat room to type text into.  here's screen-sharing.  And you can put up your hand and see stats on who talked the most.  That's it.

So there's not much to "integrate", Jitsi has no users, chances are random people not in your org without nextcloud might be on there.  The text chat tool is just meant to help you along with meeting stuff, it's really not something we need to integrate into anything.  

This different approach took some getting used to... I am used to putting more effort into security than "hey don't worry be happy".  But it actually works!  I see why people love it, that is actually a huge feature for that use case.  It replicates virtually the simplicity of real life, there is zero barrier to entry.  Talk can of course invite anonymous users too and shares that feature, but Jitsi it's taken to extremes as a way of life, there are no other kinds of users but anonymous!

so this works out great.... as long as you're not going to rack up a huge bandwidth bill if one of your employees starts up a 50 person fam-jam and runs it for months or something like that!

But otherwise, you don't need to worry about pepole hacking in etc because it's just wide open. Crazy!  But it works!!  It works so well, it's perfect for just giving a bunch of recent converts to online workflows somewhere to just simulate the old meet in a room in person.  It's not got  any of Talk's cool features or integration etc.  But it will happily scale up until the host PC croaks and it also runs great in a virtual environment.  It's easy and cheap to fire one up on a cloud server or some old piece of junk under your desk and it will easily handle an org's needs.  It also means you might as well just use a shared server from someone you trust. I am in Canada and EasyDNS who are also Canadian offer a public server.  Lots of orgs in lots of countries do, so users don't need to intsall Jitsi at all to use this, they just need to be able to point the URL at a public server they trust and they're done!

So anyway I need this... so I'm going to create said app for myself anyway within the next 2 weeks.  I will contrib it to the app store with a name like "Jitsi Launcher" or something so it's obvious it doesn't imply tight integration with NC and there is room for a proper "Jitsi" one day in the future maybe that is properly integrated (if that's even a thing?  Jitsi may get more features later..)     I'll leave a link here to that project once I get it going.

**Comment by sunjam** [comment]
Use the external sites app

**Comment by alexanderdd** [comment]
> Use the external sites app

I used this CC0 svg logo from here https://www.svgrepo.com/svg/40483/video-chat and I made the white version, also CC0. Feel free to use and distribute.
Upload the white and black svg to the external files app, it is already named in a way that it recognizes that they belong togehter.
[videochat-logos-for-external-sites-app-CC0.zip](https://github.com/nextcloud/spreed/files/5883703/videochat-logos-for-external-sites-app-CC0.zip)

**Comment by dosch** [comment]
@sunjam & @damienvancouver 
This sounds great! The only one feature I would need is the option the create a *random* room. 
Using the external site button, you always go to the same jitsi room... 

Did you @damienvancouver go ahead with this idea? How did it develop?

**Comment by sunjam** [comment]
Cool, that would be something to check on the Jitsi side. A
jitsi.com/create-new-room url that re-directs to a newly generated room.
Unsure if that already exists, but check their docs and github.

**Comment by sekarpdkt** [comment]
I tried this and hopefully, this is ok for most of our use case.

Step #1: Configure external site. URL is configured as `<Your Nextcloud domain>/apps/jitsi/?uid={uid}&email={email}&displayname={displayname}&domain=meet.jit.si`
Step #2: Used this image as icon
![jitsi](https://user-images.githubusercontent.com/2845842/115153271-ad69e600-a092-11eb-8b87-c036ccd10919.png)
Step#3: In apps directory, created a directory `jitsi`
Step#4: Placed this as `index.html` inside that directory

```
<!doctype html>
<html>
    <head>
<script src='https://meet.jit.si/external_api.js'></script>

<script>

function uuidv4() {
  return ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c =>
    (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
  );
}

function init(){

const queryString = window.location.search;
const urlParams = new URLSearchParams(queryString);
const domain = urlParams.get('domain') || 'meet.jit.si';

const options = {
    roomName: uuidv4(),
    width: window.screen.availWidth,
    height: window.screen.availHeight,
        userInfo: {
        email: urlParams.get('email'),
        displayName: urlParams.get('uid')
    },
    parentNode: document.querySelector('.meet'),

};
const api = new JitsiMeetExternalAPI(domain, options);
} 
</script>
    </head>
    <body class="meet" onload='init()'>
    </body>
</html>

```

**Comment by loxK** [comment]
My use case is creating a video chat invitation on a configurable Jitsi server the same way it can be done for Talk. With a URL attached to the event.

**Comment by sekarpdkt** [comment]
I spent few hours in past couple of days. Thought of following requirements and approach

1. Shall allow only authorised users to use Jitsi meet ==> Use hosted Jitsi with AWT authentication
2. Shall be able to create adhoc meeting ==> Tweak above to ensure it is a PHP app and create JWT token and redirect to Jitsi. PHP app shall auto pick userID and user who accessed the above html (converted to PHP) and include it in JWT. In addition, if meeting ID is provided in URL param, use it instead of generating new one.
3. Shall be able to attach Jitsi meeting invite to calendar ==> In `https://github.com/nextcloud/calendar/blob/master/src/services/talkService.js`, tweek following function

```

/**
 * Generates an absolute URL to the talk room based on the token
 *
 * @param {String} token The token to the call room
 * @returns {string}
 */
function generateURLForToken(token = '') {
	return window.location.protocol + '//' + window.location.host + generateUrl('/call/' + token)
}

```
In this js, include UUID logic and generate meetingID and point it to above step#2 URL. Which will pick up meeting ID from URL and attach JWT and redirect to hosted Jitsi. Try to search

`return window.location.protocol+"//"+window.location.host+Object(et.generateUrl)("/call/"+e)}`

in /apps/calendar/js/calendar.js and tweak from POC point of view. We may need help from maintainer of calendar to make it as permanent config. may be add another button (add custom room and allow admin to set URL pattern)?

4. Allow only those invited to join the meeting (even if they get to know the URL in step#3 ==> Is it possible to send calendar ID in step#3 URL and in PHP step#2, check the invitee list against user id and restrict 
5. Record meeting to Nextcloud ==> Not sure... Can we use RTMP streaming option and NGINX to record it in Nextcloud data directory? Some hints.

```
https://github.com/jitsi/jitsi-meet/issues/8955
https://github.com/jitsi/jibri/pull/216

https://opensource.com/article/20/2/video-streaming-tools
https://stackoverflow.com/questions/64274829/how-can-i-convert-rtmp-nginx-livestream-from-flv-to-mp4-file
https://www.servermania.com/kb/articles/nginx-rtmp/
```
